libs/aws/langchain_aws/agents/base.py

```diff
@@ -0,0 +1,266 @@
+        'name': function_name,
+        'parameters': function_parameters
+    }
+
+
+def _prepare_agent(bedrock_client, agent_id):
+    bedrock_client.prepare_agent(agentId=agent_id)
+    prepare_agent_start_time = time.time()
+    while time.time() - prepare_agent_start_time < 10:
+        agent_status = bedrock_client.get_agent(agentId=agent_id)
+        if agent_status.get('agent', {}).get('agentStatus', '') == 'PREPARED':
+            return
+        else:
+            time.sleep(2)
+    raise Exception(f'Timed out while preparing the agent with id {agent_id}')
+
+
+class BedrockAgentsRunnable(RunnableSerializable[Dict, OutputType]):
+    client: Any = Field(default_factory=get_bedrock_agents_runtime_client)
+    agent_alias_id: str = _TEST_AGENT_ALIAS_ID
+    agent_id: Optional[str] = None
+    session_id: Optional[str] = None
+    enable_trace: bool = False
+
+    @root_validator
+    def validate_agent(cls, values: dict) -> dict:
+        # Can create and prepare agents here
+        # Set agent_id, agent_alias_id, session_id etc.
+        if not values["session_id"]:
+            values["session_id"] = uuid.uuid4().hex
+
+        return values
+
+    @classmethod
+    def create_agent(cls,
+                     agent_name: str,
+                     agent_resource_role_arn: str,
+                     model: str,
+                     instructions: str,
+                     tools: List[BaseTool],
+                     agent_alias_id: str = _TEST_AGENT_ALIAS_ID,
+                     **kwargs: Any,
+                     ) -> BedrockAgentsRunnable:
+        try:
+            bedrock_client = get_bedrock_agents_client()
+            agent_id = _create_bedrock_agent(bedrock_client, agent_name, agent_resource_role_arn, instructions, model)
+            _create_bedrock_action_groups(bedrock_client, agent_id, tools)
+            _prepare_agent(bedrock_client, agent_id)
+
+            return cls(
+                agent_id=agent_id,
+                **kwargs
+            )
+        except Exception as exception:
+            logging.error(f'Error in create agent call: {exception}')
+            raise exception
+
+    def invoke(
+            self, input: Dict, config: Optional[RunnableConfig] = None
+    ) -> OutputType:
+        config = ensure_config(config)
+        agent_input = {
+            "agentId": self.agent_id,
+            "agentAliasId": self.agent_alias_id,
+            "enableTrace": self.enable_trace,
+            "sessionId": self.session_id,
+        }
+        if input.get("intermediate_steps"):
+            session_state = self._parse_intermediate_steps(
+                input.get("intermediate_steps")  # type: ignore[arg-type]
+            )
+            if session_state is not None:
+                agent_input["sessionState"] = session_state
+        else:
+            agent_input["inputText"] = input.get("input", "")
+
+        output = self.client.invoke_agent(**agent_input)
+        return parse_agent_response(output)
```

Comment:
Add callbacks related code and message handling.

Reply:
done